SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LifetimeDependenceScopeFixup.swift

```diff
@@ -40,93 +40,110 @@ let lifetimeDependenceScopeFixupPass = FunctionPass(
+    return walker.dependsOnCaller ? arg : nil
+  }
+  return nil
+}
 
-    // Delete original end_access instructions
-    for endAccess in endAcceses {
-      context.erase(instruction: endAccess)
-    }
+/// Extend this access scope to cover the dependent uses. Recursively extend outer accesses to maintain nesting.
+///
+/// Note that we cannot simply rewrite the `mark_dependence` to depend on an outer access scope. For 'read' access, this
+/// could let us avoid extending the inner scope, but that would not accomplish anything useful because inner 'read's
+/// can always be extended up to the extent of their outer 'read' (ignoring the special case when the dependence is on a
+/// caller scope, which is handled separately). A nested 'read' access can never interfere with another access in the
+/// same outer 'read', because it is impossible to nest a 'modify' access within a 'read'. For 'modify' accesses,
+/// however, the inner scope must be extended for correctness. A 'modify' access can interfere with other 'modify'
+/// accesss in the same scope. We rely on exclusivity diagnostics to report these interferences. For example:
+///
+///     sil @foo : $(@inout C) -> () {
+///       bb0(%0 : $*C):
+///         %a1 = begin_access [modify] %0
+///         %d = apply @getDependent(%a1)
+///         mark_dependence [unresolved] %d on %a1
+///         end_access %a1
+///         %a2 = begin_access [modify] %0
+///         ...
+///         end_access %a2
+///         apply @useDependent(%d) // exclusivity violation
+///         return
+///     }
+///
+/// The call to `@useDependent` is an exclusivity violation because it uses a value that depends on a 'modify'
+/// access. This scope fixup pass must extend '%a1' to cover the `@useDependent` but must not extend the base of the
+/// `mark_dependence` to the outer access `%0`. This ensures that exclusivity diagnostics correctly reports the
+/// violation, and that subsequent optimizations do not shrink the inner access `%a1`.
+private func extendAccessScope(beginAccess: BeginAccessInst, range: inout InstructionRange,
+                               _ context: FunctionPassContext) -> FunctionArgument? {
+  var endAcceses = [Instruction]()
+  // Collect the original end_access instructions and extend the range to to cover them. The access scope should only be
+  // extended here; it may protect other memory operations.
+  for end in beginAccess.endInstructions {
+    endAcceses.append(end)
+    range.insert(end)
+  }
+  assert(!range.ends.isEmpty)
 
-    // TODO: Add SIL support for lifetime dependence and write unit test
-    // for nested access scopes
-    guard case let .scope(enclosingBeginAccess) = beginAccess.address.enclosingAccessScope else {
-      break
+  // Create new end_access at the end of extended uses
+  for end in range.ends {
+    let endBuilder = Builder(after: end, context)
+    _ = endBuilder.createEndAccess(beginAccess: beginAccess)
```

Comment:
`Builder.createEndAccess` should have a `@discardableResult` attribute, because `end_access` has no result value.